File: app/src/main/java/com/test/tclick/Func.java
package com.test.tclick;

import android.accessibilityservice.AccessibilityService;
import android.accessibilityservice.GestureDescription;
import android.graphics.Path;
import android.graphics.Rect;
import android.os.Build;
import android.os.Bundle;
import android.os.SystemClock;
import android.util.Log;
import android.view.MotionEvent;
import android.view.accessibility.AccessibilityNodeInfo;

import java.util.List;

public class Func {
    public AccessibilityService service;

    public Func(AccessibilityService service) {
        this.service = service;
    }

    public int[] getWindowSize() {
        return new int[]{
                service.getResources().getDisplayMetrics().widthPixels,
                service.getResources().getDisplayMetrics().heightPixels
        };
    }

    public void scrollMove(String move, int num, int back) {
        int[] size = getWindowSize();
        int width = size[0];
        int height = size[1];

        if ("up".equals(move)) {
            for (int i = 0; i < num; i++) {
                swipe(width / 2, height / 2, width / 2, height / 4);
                SystemClock.sleep(300);
            }
            Log.d("Func", "스크롤 " + move + " " + num + " 번");
        } else if ("down".equals(move)) {
            for (int i = 0; i < num; i++) {
                swipe(width / 2, height / 4, width / 2, height / 2);
                SystemClock.sleep(300);
            }
            Log.d("Func", "스크롤 " + move + " " + num + " 번");
        } else {
            for (int i = 0; i < num; i++) {
                swipe(width / 2, height / 2, width / 2, height / 4);
                swipe(width / 2, height / 4, width / 2, height / 2);
            }
            for (int i = 0; i < back; i++) {
                service.performGlobalAction(AccessibilityService.GLOBAL_ACTION_BACK);
            }
            Log.d("Func", "스크롤 " + move + " " + num + " 번");
            Log.d("Func", "뒤로가기 " + back + " 번");
        }
    }

    private void swipe(int startX, int startY, int endX, int endY) {
        // Gesture를 사용하여 스와이프 구현이 필요합니다.
        // Android 7.0 (API 레벨 24) 이상에서 `GestureDescription` 클래스를 사용하여 구현할 수 있습니다.
    }

    public AccessibilityNodeInfo findElementByText(String text) {
        AccessibilityNodeInfo rootNode = service.getRootInActiveWindow();
        if (rootNode != null) {
            List<AccessibilityNodeInfo> nodes = rootNode.findAccessibilityNodeInfosByText(text);
            if (!nodes.isEmpty()) {
                return nodes.get(0);
            }
        }
        return null;
    }

    public AccessibilityNodeInfo findElementByTextContains(String text) {
        AccessibilityNodeInfo rootNode = service.getRootInActiveWindow();
        if (rootNode != null) {
            List<AccessibilityNodeInfo> nodes = rootNode.findAccessibilityNodeInfosByText(text);
            if (!nodes.isEmpty()) {
                for (AccessibilityNodeInfo node : nodes) {
                    if (node.getText() != null && node.getText().toString().contains(text)) {
                        return node;
                    }
                }
            }
        }
        return null;
    }

    public AccessibilityNodeInfo findElementByViewId(String viewId) {
        AccessibilityNodeInfo rootNode = service.getRootInActiveWindow();
        if (rootNode != null) {
            List<AccessibilityNodeInfo> nodes = rootNode.findAccessibilityNodeInfosByViewId(viewId);
            if (!nodes.isEmpty()) {
                return nodes.get(0);
            }
        }
        return null;
    }

    private AccessibilityNodeInfo findSibling(AccessibilityNodeInfo parentElement, String siblingClass) {
        if (parentElement != null && parentElement.getParent() != null) {
            AccessibilityNodeInfo parent = parentElement.getParent();
            int childCount = parent.getChildCount();
            for (int i = 0; i < childCount; i++) {
                AccessibilityNodeInfo sibling = parent.getChild(i);
                if (sibling != null && sibling.getClassName().equals(siblingClass)) {
                    return sibling;
                }
            }
        }
        return null;
    }

    private AccessibilityNodeInfo findChild(AccessibilityNodeInfo siblingElement, String childClass) {
        if (siblingElement != null) {
            int childCount = siblingElement.getChildCount();
            for (int i = 0; i < childCount; i++) {
                AccessibilityNodeInfo child = siblingElement.getChild(i);
                if (child != null && child.getClassName().equals(childClass)) {
                    return child;
                }
            }
        }
        return null;
    }

    private void clickElement(AccessibilityNodeInfo element) {
        if (element != null) {
            element.performAction(AccessibilityNodeInfo.ACTION_CLICK);
        }
    }

    private void longClickElement(AccessibilityNodeInfo element) {
        if (element != null) {
            element.performAction(AccessibilityNodeInfo.ACTION_LONG_CLICK);
        }
    }

    public boolean clickElement(AccessibilityNodeInfo element, String text, long sleepTime, int clicks, boolean longClick) {
        if (element != null) {
            for (int i = 0; i < clicks; i++) {
                if (longClick) {
                    longClickElement(element);
                } else {
                    clickElement(element);
                }
                Log.d("Func", text + " click");
                SystemClock.sleep(sleepTime);
            }
            return true;
        } else {
            Log.d("Func", "clickElement 함수에서 not found " + text + " element");
            return false;
        }
    }

    public boolean clickElement2(String text, long sleepTime, int clicks, String... args) {
        AccessibilityNodeInfo element = findElementByText(text);
        if (element != null) {
            for (int i = 0; i < clicks; i++) {
                clickElement(element);
                Log.d("Func", text + " 클릭");
                SystemClock.sleep(sleepTime);
            }
            return true;
        }
        return false;
    }

    public boolean findAndClick(String text, boolean contains, long sleepTime, int clicks, boolean longClick) {
        long startTime = System.currentTimeMillis();
        long timeout = 10000; // 최대 10초 동안 대기
        boolean timeoutReached = false; // 타임아웃 도달 여부

        while (true) {
            boolean isLoading = checkLoading();
            Log.d("Func", "checkLoading 결과: " + isLoading);
            if (isLoading) {
                SystemClock.sleep(500);
                break;
            } else {
                SystemClock.sleep(1000);
            }

            // 타임아웃 체크
            if (System.currentTimeMillis() - startTime > timeout) {
                Log.d("Func", "타임아웃 도달. 반복 종료");
                timeoutReached = true;
                break;
            }
        }

        // 타임아웃 도달 시 false 반환하고 종료
        if (timeoutReached) {
            return false;
        }

        AccessibilityNodeInfo element = contains ? findElementByTextContains(text) : findElementByText(text);
        if (element != null) {
            String elementText = getElementText(element);
            Log.d("Func", "Element 발견: " + elementText);

            if (isInvalidElement(element, elementText)) {
                Log.d("Func", "Element가 유효하지 않음: " + elementText);
                return false;
            } else {
                boolean clicked = clickElement(element, text, sleepTime, clicks, longClick);
                Log.d("Func", "Element 클릭 결과: " + clicked);
                return clicked;
            }
        } else {
            Log.d("Func", "findAndClick 함수에서 not found: " + text);
            return false;
        }
    }



    public void mainBack(String text, String text2, boolean contains) {
        long startTime = System.currentTimeMillis();
        long timeout = 3000;  // 최대 3초 동안 시도

        for (int i = 0; i < 10; i++) {
            SystemClock.sleep(300);
            service.performGlobalAction(AccessibilityService.GLOBAL_ACTION_BACK);
            Log.d("Func", "뒤로가기 시도 " + (i + 1));

            if (!text2.isEmpty()) {
                boolean clicked = findAndClick(text2, false, 500, 1, false);
                if (clicked) {
                    Log.d("Func", text2 + " 클릭 성공");
                } else {
                    Log.d("Func", text2 + " 클릭 실패");
                }
            }

            boolean found = (contains && findElementByTextContains(text) != null) ||
                    (!contains && findElementByText(text) != null);

            if (found) {
                Log.d("Func", text + " 발견됨");
                break;
            }

            // 타임아웃 체크
            if (System.currentTimeMillis() - startTime > timeout) {
                Log.d("Func", "타임아웃 도달. 반복 종료");
                break;
            }
        }
    }

    public void backKey(int num2) {
        for (int i = 0; i < num2; i++) {
            service.performGlobalAction(AccessibilityService.GLOBAL_ACTION_BACK);
            SystemClock.sleep(300);
        }
    }

    public void waitElement(boolean existElement, String... args) {
        for (int i = 0; i < 20; i++) {
            if (findElementByTextContains(args[0]) != null == existElement) {
                break;
            }
            SystemClock.sleep(300);
        }
    }

    private boolean checkLoading() {
        AccessibilityNodeInfo loading = findElementByTextContains("Loading");
        return loading != null;
    }

    private String getElementText(AccessibilityNodeInfo element) {
        CharSequence text = element.getText();
        return text != null ? text.toString() : "";
    }

    private boolean isInvalidElement(AccessibilityNodeInfo element, String text) {
        return text.equals("") || element.getContentDescription() != null;
    }

    public void deviceClick(int x, int y) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            // API 24 이상에서 GestureDescription 사용
            GestureDescription.Builder gestureBuilder = new GestureDescription.Builder();
            Path path = new Path();
            path.moveTo(x, y);
            gestureBuilder.addStroke(new GestureDescription.StrokeDescription(path, 0, 1));
            GestureDescription gesture = gestureBuilder.build();
            service.dispatchGesture(gesture, null, null);
        } else {
            // API 24 이하에서는 MotionEvent 사용
            MotionEvent eventDown = MotionEvent.obtain(SystemClock.uptimeMillis(), SystemClock.uptimeMillis(),
                    MotionEvent.ACTION_DOWN, x, y, 0);
            MotionEvent eventUp = MotionEvent.obtain(SystemClock.uptimeMillis(), SystemClock.uptimeMillis() + 100,
                    MotionEvent.ACTION_UP, x, y, 0);
//            service.dispatchTouchEvent(eventDown);
//            service.dispatchTouchEvent(eventUp);
            eventDown.recycle();
            eventUp.recycle();
        }
    }

    // 텍스트 필드를 찾아서 마지막 4자를 제외한 나머지 부분을 삭제하는 함수
    public void deleteTextExceptLast4(AccessibilityNodeInfo node) {
        if (node != null) {
            // 텍스트 필드 클릭
            node.performAction(AccessibilityNodeInfo.ACTION_FOCUS);

            // 전체 텍스트 선택
            CharSequence text = node.getText();
            if (text != null) {
                String textString = text.toString();
                int textLength = textString.length();

                if (textLength > 4) {
                    // 마지막 4자를 제외한 텍스트 선택
                    int start = 0;
                    int end = textLength - 4;
                    Bundle arguments = new Bundle();
                    arguments.putInt(AccessibilityNodeInfo.ACTION_ARGUMENT_SELECTION_START_INT, start);
                    arguments.putInt(AccessibilityNodeInfo.ACTION_ARGUMENT_SELECTION_END_INT, end);
                    node.performAction(AccessibilityNodeInfo.ACTION_SET_TEXT, arguments);

                    // 선택한 텍스트 삭제
                    arguments.putCharSequence(AccessibilityNodeInfo.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE, "");
                    node.performAction(AccessibilityNodeInfo.ACTION_SET_TEXT, arguments);
                } else {
                    // 텍스트 길이가 4보다 작을 경우 전체 텍스트를 삭제
                    Bundle arguments = new Bundle();
                    arguments.putCharSequence(AccessibilityNodeInfo.ACTION_ARGUMENT_SET_TEXT_CHARSEQUENCE, "");
                    node.performAction(AccessibilityNodeInfo.ACTION_SET_TEXT, arguments);
                }
            }
        }
    }
}

File: app/src/main/java/com/test/tclick/Func.kt
package com.test.tclick

import android.accessibilityservice.AccessibilityService
import android.content.Context
import android.content.Context.WINDOW_SERVICE
import android.graphics.Rect
import android.os.SystemClock
import android.util.DisplayMetrics
import android.util.Log
import android.view.WindowManager
import android.view.accessibility.AccessibilityNodeInfo
import kotlin.math.roundToInt

class Func(private val service: AccessibilityService) {

    fun findTextElements(rootNode: AccessibilityNodeInfo) {
        // rootNode가 null인 경우 함수 종료
        if (rootNode == null) return

        // 찾고자 하는 텍스트 목록
        val textsToFind = listOf("혜택", "혜택", "친구와", "1원", "보너스")

        // 각 텍스트에 대해 탐색
        for (text in textsToFind) {
            val nodes = rootNode.findAccessibilityNodeInfosByText(text)
            if (nodes.isNotEmpty()) {
                Log.d("MyAccessibilityService", "Found text: $text")

                // 첫 번째 노드 탐색
                nodes.firstOrNull()?.let { node ->
                    if (node.isClickable) {
                        node.performAction(AccessibilityNodeInfo.ACTION_CLICK)
                        Log.d("MyAccessibilityService", "Clicked text: $text")
                    } else {
                        // 부모 노드 중 클릭 가능한 노드 찾기
                        var parent = node.parent
                        while (parent != null) {
                            if (parent.isClickable) {
                                parent.performAction(AccessibilityNodeInfo.ACTION_CLICK)
                                Log.d("MyAccessibilityService", "Clicked parent node for text: $text")
                                break
                            }
                            parent = parent.parent
                        }
                        if (parent == null) {
                            Log.d("MyAccessibilityService", "No clickable parent found for text: $text")
                        }
                    }
                    SystemClock.sleep(1000) // 1초 대기
                }
            } else {
                Log.d("MyAccessibilityService", "Text not found: $text")
            }
        }
    }

    fun mainBack1(
        service: AccessibilityService,  // AccessibilityService 인스턴스를 받음
        rootNode: AccessibilityNodeInfo,
        targetText: String = "",  // 찾아야 할 텍스트
        sleepTime: Long = 300,    // 버튼 클릭 후 대기 시간
        maxAttempts: Int = 25     // 최대 시도 횟수
    ) {
        if (rootNode == null) return
        var attemptCount = 0
        var textFound = false

        while (attemptCount < maxAttempts) {
            // targetText가 화면에 있는지 확인
            val nodes = rootNode.findAccessibilityNodeInfosByText(targetText)
            if (nodes.isNotEmpty()) {
                Log.d("Func", "$targetText found, stopping back press at attempt $attemptCount")
                // targetText를 찾으면 작업을 멈춤
                return
            } else {
                // targetText가 화면에 없으면 백 버튼 누르기
                service.performGlobalAction(AccessibilityService.GLOBAL_ACTION_BACK)
                SystemClock.sleep(sleepTime)

                attemptCount++
                Log.d("Func", "Performed back press $attemptCount time(s)")
            }
        }

        Log.d("Func", "Exceeded maximum attempts of $maxAttempts without finding $targetText")
    }

    fun getScreenSize(service: AccessibilityService): Pair<Int, Int>? {
        val displayMetrics = service.resources.displayMetrics
        val width = displayMetrics.widthPixels
        val height = displayMetrics.heightPixels
        return Pair(width, height)
    }

    fun scrollMove(service: AccessibilityService, rootNode: AccessibilityNodeInfo?, move: String = "up", num: Int = 1, back: Int = 0) {
        val (width, height) = getScreenSize(service) ?: return
        for (i in 1..num) {
            when (move) {
                "up" -> {
                    Log.d("Func", "Scrolling up $num times")
                    rootNode?.performAction(AccessibilityNodeInfo.ACTION_SCROLL_BACKWARD)
                    SystemClock.sleep(500)
                }
                "down" -> {
                    Log.d("Func", "Scrolling down $num times")
                    rootNode?.performAction(AccessibilityNodeInfo.ACTION_SCROLL_FORWARD)
                    SystemClock.sleep(500)
                }
                else -> {
                    Log.d("Func", "Unknown scroll direction")
                }
            }
        }
        for (i in 1..back) {
            Log.d("Func", "Pressing back button $back times")
            service.performGlobalAction(AccessibilityService.GLOBAL_ACTION_BACK)
            SystemClock.sleep(500)
        }
    }

    fun clickElement(element: AccessibilityNodeInfo?, text: String, sleepTime: Long = 500, clicks: Int = 1, longClick: Boolean = false): Boolean {
        element?.let {
            for (i in 1..clicks) {
                if (longClick) {
                    Log.d("Func", "Long clicking on $text")
                    it.performAction(AccessibilityNodeInfo.ACTION_LONG_CLICK)
                } else {
                    Log.d("Func", "Clicking on $text")
                    it.performAction(AccessibilityNodeInfo.ACTION_CLICK)
                }
                SystemClock.sleep(sleepTime)
            }
            return true
        }
        Log.d("Func", "Element $text not found")
        return false
    }

    fun findAndClick(service: AccessibilityService, rootNode: AccessibilityNodeInfo?, text: String, contains: Boolean = false, sleepTime: Long = 500, clicks: Int = 1, longClick: Boolean = false): Boolean {
        rootNode?.let {
            val elements = if (contains) {
                it.findAccessibilityNodeInfosByText(text)
            } else {
                it.findAccessibilityNodeInfosByViewId(text)
            }

            elements?.let { nodes ->
                nodes.forEach { node ->
                    if (node.text.contains("완료") || node.text.contains("후에")) {
                        Log.d("Func", "Skipping '완료' or '후에'")
                        return false
                    }
                    return clickElement(node, text, sleepTime, clicks, longClick)
                }
            }
        }
        Log.d("Func", "Text $text not found")
        return false
    }

    fun mainBack(service: AccessibilityService, rootNode: AccessibilityNodeInfo,targetText: String = "", text2: String = "", sleepTime: Long = 1000) {
        var count = 0
        while (count < 25) {
            SystemClock.sleep(sleepTime)
            if (text2.isNotEmpty()) {
                findAndClick(service, rootNode, text2)
            }
            val nodes = rootNode.findAccessibilityNodeInfosByText(targetText)
            if (nodes.isNotEmpty()) {
                Log.d("Func", "$targetText found, stopping back press at attempt $count")
                // targetText를 찾으면 작업을 멈춤
                return
            } else {
                // targetText가 화면에 없으면 백 버튼 누르기
                service.performGlobalAction(AccessibilityService.GLOBAL_ACTION_BACK)
                SystemClock.sleep(sleepTime)

                count++
                Log.d("Func", "Performed back press $count time(s)")
            }

            Log.d("Func", "$count time back")
            Log.d("Func", "Exceeded maximum attempts of $count without finding $targetText")
        }
    }

    fun backKey(service: AccessibilityService, num: Int = 1) {
        for (i in 1..num) {
            Log.d("Func", "Pressing back key")
            service.performGlobalAction(AccessibilityService.GLOBAL_ACTION_BACK)
            SystemClock.sleep(300)
        }
    }

    fun waitElement(service: AccessibilityService, exist: Boolean = true, rootNode: AccessibilityNodeInfo?, text: String) {
        for (i in 1..8) {
            if (rootNode != null) {
                if (exist) {
                    val elementExists = rootNode.findAccessibilityNodeInfosByText(text).isNotEmpty()
                    if (elementExists) {
                        Log.d("Func", "Element $text found")
                        SystemClock.sleep(1000)
                        return  // break 대신에 return으로 함수 자체를 종료
                    } else {
                        Log.d("Func", "Waiting for $text element to appear...")
                        SystemClock.sleep(2000)
                    }
                } else {
                    val elementExists = rootNode.findAccessibilityNodeInfosByText(text).isEmpty()
                    if (elementExists) {
                        Log.d("Func", "Waiting for $text element to disappear...")
                        return  // break 대신에 return으로 함수 자체를 종료
                    } else {
                        Log.d("Func", "$text element still exists...")
                        SystemClock.sleep(2000)
                    }
                }
            }
        }
    }


    fun findAndClickSiblingChild(service: AccessibilityService, rootNode: AccessibilityNodeInfo?, parentText: String, siblingClass: String, childClass: String, contains: Boolean = false, sleepTime: Long = 500, clicks: Int = 1): Boolean {
        val parentElement = if (contains) {
            rootNode?.findAccessibilityNodeInfosByText(parentText)
        } else {
            rootNode?.findAccessibilityNodeInfosByViewId(parentText)
        }

        parentElement?.let { nodes ->
            nodes.forEach { parent ->
                val siblingElement = parent.getChild(0) // Assume sibling is the first child. Adjust if needed.
                siblingElement?.let { sibling ->
                    val childElement = sibling.getChild(0) // Adjust index accordingly.
                    if (childElement != null) {
                        return clickElement(childElement, childClass, sleepTime, clicks)
                    } else {
                        Log.d("Func", "Child element not found")
                    }
                } ?: run {
                    Log.d("Func", "Sibling element not found")
                }
            }
        }
        Log.d("Func", "Parent element with text $parentText not found")
        return false
    }

    fun findAndClickWithinBounds(service: AccessibilityService, textContains: String, endText: String, rootNode: AccessibilityNodeInfo?, interval: Int = 120, startX: Int = 200, startY: Int = 740) {
        rootNode?.let {
            val elements = it.findAccessibilityNodeInfosByText(textContains)
            elements?.forEach { element ->
                val bounds = Rect() // Rect 객체 생성
                element.getBoundsInScreen(bounds) // 경계를 가져오는 메소드 사용
                Log.d("Func", "Element bounds: $bounds")

                // 경계를 기반으로 클릭 좌표 설정
                for (offset in 0 until bounds.bottom - bounds.top step interval) {
                    val clickX = startX + offset
                    val clickY = startY
                    Log.d("Func", "Clicking at ($clickX, $clickY)")

                    // 해당 요소에 대해 클릭 동작 수행
                    element.performAction(AccessibilityNodeInfo.ACTION_CLICK)
                    SystemClock.sleep(500)

                    // 화면 전환 감지
                    if (it.findAccessibilityNodeInfosByText(endText).isEmpty()) {
                        Log.d("Func", "Screen transition detected after click")
                        break
                    }
                }
            }
        }
    }
}
